### DISCUSSION TITLE
How can I store the state a database?

### DISCUSSION THREAD

**Discussion by afagard** [spatie/laravel-livewire-wizard#110] [Q&A]
I want the ability to store a form wizards state in the database. I think this is really important for multi-step wizards as people may leave the page on accident. 

On first load I check to see if a user has a previous wizard via a cookie:

1. If not, then when `submit()` is called _once the user completes the first step_ I create the wizard record and save it's ID to a cookie
2. If yes, I set the initial state using the database entry

At each step submit, before `nextStep()` is called. I save the state.

The problem I am facing is when 1 happens above and I want to be able to access the database entry ID without attempting to get the cookie in subsequent steps. How can I store the entry ID to the global state? I've tried:

```
class QuestionnaireComponent extends WizardComponent
{
    use QuestionnaireMounterTrait;

    public array $allStepState = [];

    public Questionnaire|false $questionnaire;

    public function __construct()
    {
        $this->questionnaire = $this->getQuestionnaireModel(true);
        if ($this->questionnaire) {
            $this->allStepState['qid'] = $this->questionnaire->id;
            $this->currentStepName = $this->questionnaire->current_step;
        }
    }
```

which works when the user already has a database entry, however when the user doesn't and this code triggers on the first step:

```
    public function submit(): void
    {
        $this->validate();

        // this is the first run of the questionnaire for this user
        // or the users previous questionnaire is invalid
        // or the user blocks cookies
        if (empty($this->allStepsState['qid'])) {
            $this->allStepsState['qid'] = $this->getQuestionnaireModel()->id;
        }

        Questionnaire::findOrFail($this->state()->get('qid'))->fill([
            'authorization' => true,
            'current_step' => $this->allStepNames[1] // iterate to the next step
        ])->save();

        $this->nextStep();
    }
```

a dd of `$this->state()->get('qid')` before `nextStep()` is called shows the ID persisting, but when I reach the next step the ID is not there.

@freekmurze I would really appreciate it if you could take a look, your packages have greatly helped me over the years so I hate to ask more of you, but I've been wracking my brain on this one.

[QuestionnaireComponent.txt](https://github.com/spatie/laravel-livewire-wizard/files/14732017/QuestionnaireComponent.txt)
[QuestionnaireMounterTrait.txt](https://github.com/spatie/laravel-livewire-wizard/files/14732022/QuestionnaireMounterTrait.txt)
[AuthorizationStepComponent.txt](https://github.com/spatie/laravel-livewire-wizard/files/14732023/AuthorizationStepComponent.txt)

**Comment by ju5t** [comment]
You probably need to use an initial state. See https://spatie.be/docs/laravel-livewire-wizard/v2/usage/setting-initial-state. This will allow you to fetch the state from the database and make it available in all steps.

**Reply by afagard** [reply]
Indeed, I am aware of that, however assuming this is the first wizard run for this user, and to prevent unnecessary database entries on page-load, I am only creating the record once step 1 is finished at which point I need to somehow merge the ID in to the "global state". Let me know if I didn't explain that adequately above :)

When the database entry already exists on page-load, then I do utilize the initial state to populate all the properties of the various steps. That part works.

**Reply by ju5t** [reply]
Eh, not sure if I understand you.

Is it only the `$this->getQuestionnaireModel()->id` you want to store globally? If so, I would probably use a session for that.

**Reply by afagard** [reply]
I thought about that, but it seems as though each "step" is able to be merged in to the global state of the component because we can use `$this->state()->all()` to get all the state of other components. I guess my confusion is why we can't do that with a simple ID.

So for clarity, when my wizard component starts I run getQuestionnaireModel(false) and then if it returns a valid model I start setting the steps initial state in addition I assign 

Wizard constructor (see .txt files in original question):
(this works and qid can be accessed in $this->state()->get('qid') throughout the steps)
```
        $this->questionnaire = $this->getQuestionnaireModel(true);
        if ($this->questionnaire) {
            $this->allStepState['qid'] = $this->questionnaire->id;
```
 If it doesn't return a valid model, then I just create a model after the user completes the first step and store that in a cookie:
 
First step component - this doesn't persist to the next step :(
```
         if (empty($this->allStepsState['qid'])) {
            $this->allStepsState['qid'] = $this->getQuestionnaireModel()->id;
        }
```

Note: since we last spoke I added more code to the above question.

**Reply by afagard** [reply]
Edit to the above: I think it has something to do with component nesting. When the qid is set on the parent level (the wizard rather than the step) it persists, but when it is set from the child it doesn't which I suppose makes sense. But ultimately I just need it to be added to the state like the rest of the items available in $this->state()->all()
